File: CpmServer/cpm-web/src/api/pmProjectTrace.ts
import request from './request'

export interface ApiResponse<T> {
  code: number
  message: string
  data: T
}

export interface PmProjectTraceStepActualCycleTime {
  id?: number
  projectTraceStepId?: number
  recordDate: string
  actualCycleTime?: number
  remarks?: string
  status?: number
}

export interface PmProjectTraceStep {
  id?: number
  stepOrder: number
  processName: string
  personInCharge?: string
  cycleTime?: number
  settingDays?: number
  estimatedHours?: number
  remarks?: string
  planDurationDays?: number
  planStartDate?: string
  planEndDate?: string
  actualStartDate?: string
  actualForecastStartDate?: string
  actualDurationDays?: number
  actualPlanDurationDays?: number
  actualEndDate?: string
  actualCycleTimes: PmProjectTraceStepActualCycleTime[]
}

export interface PmProjectTrace {
  id?: number
  quotationId?: number
  customerId: number
  customerName: string
  productId?: number
  productCode: string
  productName?: string
  plannedQty?: number
  projectStartDate?: string
  displayWeeks?: number
  status: number
  createdAt?: string
  updatedAt?: string
  steps: PmProjectTraceStep[]
}

export interface TraceListResult {
  total: number
  list: PmProjectTrace[]
}

export function getTraceList(params: {
  keyword?: string
  status?: number
  page?: number
  pageSize?: number
}): Promise<ApiResponse<TraceListResult>> {
  return request.get('/PmProjectTrace', { params })
}

export function getTraceDetail(id: number): Promise<ApiResponse<PmProjectTrace>> {
  return request.get(`/PmProjectTrace/${id}`)
}

export function createTrace(data: PmProjectTrace): Promise<ApiResponse<number>> {
  return request.post('/PmProjectTrace', data)
}

export function updateTrace(id: number, data: PmProjectTrace): Promise<ApiResponse<unknown>> {
  return request.put(`/PmProjectTrace/${id}`, data)
}

export function deleteTrace(id: number): Promise<ApiResponse<unknown>> {
  return request.delete(`/PmProjectTrace/${id}`)
}

export interface PmProjectTraceStepCycleTimeItem {
  traceId: number
  customerName: string
  productCode: string
  productName?: string
  stepId: number
  stepOrder: number
  processName: string
  personInCharge?: string
  cycleTime?: number
  latestActualCycleTime?: number
  latestRecordDate?: string
  pendingRequestCount: number
}

export interface PmStepCycleTimeChangeDetail {
  changeType: number
  targetRecordId?: number
  recordDate: string
  actualCycleTime?: number
  remarks?: string
}

export function getAllStepsActualCycleTime(params: { keyword?: string }): Promise<ApiResponse<PmProjectTraceStepCycleTimeItem[]>> {
  return request.get('/PmProjectTrace/steps/actual-cycle-time', { params })
}

export function submitCycleTimeChangeRequest(data: {
  stepId: number
  traceId: number
  changes: PmStepCycleTimeChangeDetail[]
}): Promise<ApiResponse<number>> {
  return request.post('/PmProjectTrace/steps/actual-cycle-time/change-request', data)
}

File: CpmServer.Core/Models/PmProjectTrace.cs
using System.ComponentModel.DataAnnotations;
using System.ComponentModel.DataAnnotations.Schema;

namespace CpmServer.Models;

/// <summary>
/// 产品跟踪主表 - 报价单审批通过后自动生成
/// </summary>
[Table("PmProjectTrace")]
public class PmProjectTrace
{
    [Key]
    public long Id { get; set; }

    /// <summary>关联报价单</summary>
    public long? QuotationId { get; set; }

    /// <summary>客户ID（自动带入）</summary>
    public long CustomerId { get; set; }

    /// <summary>客户名称（自动带入）</summary>
    public string CustomerName { get; set; } = string.Empty;

    /// <summary>产品ID（自动带入）</summary>
    public long? ProductId { get; set; }

    /// <summary>零件号/产品编码（自动带入）</summary>
    public string ProductCode { get; set; } = string.Empty;

    /// <summary>产品名称（自动带入）</summary>
    public string? ProductName { get; set; }

    /// <summary>计划投入数量（手工填写）</summary>
    public int? PlannedQty { get; set; }

    /// <summary>项目开始日期（手工填写）</summary>
    public DateTime? ProjectStartDate { get; set; }

    /// <summary>显示周数（手工填写）</summary>
    public int? DisplayWeeks { get; set; }

    /// <summary>状态: 0=草稿 1=进行中 2=已完成</summary>
    public int Status { get; set; } = 0;

    public string? Site { get; set; }
    public string? App { get; set; }

    public DateTime CreatedAt { get; set; } = DateTime.Now;
    public DateTime UpdatedAt { get; set; } = DateTime.Now;

    public List<PmProjectTraceStep> Steps { get; set; } = new();

    [ForeignKey("QuotationId")]
    public QuoQuotation? Quotation { get; set; }

    [ForeignKey("CustomerId")]
    public CrmCustomer? Customer { get; set; }

    [ForeignKey("ProductId")]
    public CrmProduct? Product { get; set; }
}
